test/Microsoft.AspNetCore.SignalR.Redis.Tests/Docker.cs

```diff
@@ -97,28 +103,47 @@ public void Stop(ILogger logger)
             RunProcessAndThrowIfFailed(_path, $"logs {_dockerContainerName}", logger, TimeSpan.FromSeconds(5));
 
             logger.LogInformation("Stopping docker container");
-            RunProcessAndThrowIfFailed(_path, $"stop {_dockerContainerName}", logger, TimeSpan.FromSeconds(5));
+            RunProcessAndWait(_path, $"stop {_dockerMonitorContainerName}", logger, TimeSpan.FromSeconds(15), out var _);
+            RunProcessAndWait(_path, $"stop {_dockerContainerName}", logger, TimeSpan.FromSeconds(15), out var _);
         }
 
         public int RunCommand(string commandAndArguments, out string output) =>
             RunCommand(commandAndArguments, NullLogger.Instance, out output);
 
         public int RunCommand(string commandAndArguments, ILogger logger, out string output)
         {
-            return RunProcess(_path, commandAndArguments, logger, TimeSpan.FromSeconds(5), out output);
+            return RunProcessAndWait(_path, commandAndArguments, logger, TimeSpan.FromSeconds(5), out output);
         }
 
         private static void RunProcessAndThrowIfFailed(string fileName, string arguments, ILogger logger, TimeSpan timeout)
         {
-            var exitCode = RunProcess(fileName, arguments, logger, timeout, out var output);
+            var exitCode = RunProcessAndWait(fileName, arguments, logger, timeout, out var output);
 
             if (exitCode != 0)
             {
                 throw new Exception($"Command '{fileName} {arguments}' failed with exit code '{exitCode}'. Output:{Environment.NewLine}{output}");
             }
         }
 
-        private static int RunProcess(string fileName, string arguments, ILogger logger, TimeSpan timeout, out string output)
+        private static int RunProcessAndWait(string fileName, string arguments, ILogger logger, TimeSpan timeout, out string output)
+        {
+            var (process, lines) = RunProcess(fileName, arguments, logger);
+
+            if (!process.WaitForExit((int)timeout.TotalMilliseconds))
+            {
+                process.Close();
+                logger.LogError("Closing process '{processName}' because it is running longer than the configured timeout.", fileName);
+            }
+
+            // Need to WaitForExit without a timeout to guarantee the output stream has written everything
+            process.WaitForExit();
+
+            output = string.Join(Environment.NewLine, lines);
+
+            return process.ExitCode;
+        }
+
+        private static (Process, ConcurrentQueue<string>) RunProcess(string fileName, string arguments, ILogger logger)
```

Comment:
Can you put a prefix into the log messages indicating where they come from? Might need to pass that in here (since the file name will be long and awkward). I'd like to know which lines come from redis and which from the monitor process.

Reply:
Yeah, I was just thinking about that